File: w09/d04/amy_ruan/Candy_Machine/app/models/machine.rb
class Machine < ActiveRecord::Base
  attr_accessible :candy, :candy_price, :location, :money, :owner_id
  belongs_to :owner

  validates :money, numericality: true
  validates :location, presence: true
  after_initialize :set_default
  CANDY = 0
  MONEY = 0.00
  CANDY_PRICE = 0.25

  def set_default
    if self.candy.nil?
      self.candy = Machine::CANDY
    end

    if self.money.nil?
      self.money = Machine::MONEY
    end

    if self.candy_price.nil?
      self.candy_price = Machine::CANDY_PRICE
    end
  end

  def sell_candy(num)
    max_num = (num > self.candy)? self.candy : num
    money_gained = self.candy_price * max_num
    self.money += money_gained
    self.candy -= max_num
  end
end

File: w09/d04/amy_ruan/Candy_Machine/app/assets/javascripts/machine.js
Owner = Backbone.Model.extend({
  defaults: {
    money: 1000,
    candy: 7000
  },
  initialize: function(){
    alert("Hi Owner");
  }
});

var owner = new Owner({});
var money = owner.get("money");
var candy = owner.get("candy");


Machine = Backbone.Model.extend({
  defaults:{
    candy: 0,
    money: 0.00,
    candy_price: 0.25
  },
  initialize: function(){
    this.on("change:money"),
    this.on("change:money", function(model){
    var money = model.get("money");
    alert("Money: " + money);
    });
  },
  sell_candy: function(num){
    var myCandy = this.get("candy");
    var maxNum;
    if(num < myCandy){
      maxNum = num;
    }
    else{
      maxNum = myCandy;
    }
    var moneyGained = this.get("candy_price")*maxNum;
    var currentMoney = this.get("money");
    var totalMoney = moneyGained + currentMoney;
    this.set({ money: totalMoney});
    var totalCandy = myCandy-maxNum;
    this.set({ candy: totalCandy});
  }
});
